### PR TITLE
Refactor JoinExpressionBase & Improve uniquification of table alias

### PR DESCRIPTION
**Pull request by tuespetre**
This is another breakout from #7543 that will shave 9 files or so off of that PR. Summary:

- Added the `LeftJoinLateralExpression`, although it is not used yet in this PR
- Renamed `LateralJoinExpression` to `CrossJoinLateralExpression` for consistency and to resolve ambiguity
- Introduced a new base class, `PredicateJoinExpressionBase`, for `InnerJoinExpression` and `LeftOuterJoinExpression` to inherit from. This allows `CrossJoinLateralExpression` and `LeftJoinLateralExpression` to then inherit from `JoinExpressionBase`, creating new and stronger possibilities for algorithms like `HandlesQuerySource` and `GetTableForQuerySource`
- Most places using `SelectExpression.AddTable(TableExpressionBase, bool)` were passing `false`, so changed the couple of places that were not to generate the unique alias themselves in order to simplify things. One places was already doing this, resulting in unnecessary uniquification anyways.
- Instead of adding more `AddWhateverJoin` methods to `SelectExpression` for the `LeftJoinLateral`, removed all of the `AddWhateverJoin` methods and added an `AddToProjection(IEnumerable<Expression>)` method.

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftJoinLateralExpression.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Query.Sql;
+using Microsoft.EntityFrameworkCore.Utilities;
+
+namespace Microsoft.EntityFrameworkCore.Query.Expressions
+{
+    /// <summary>
+    ///     Represents a SQL LEFT JOIN LATERAL expression.
+    /// </summary>
+    public class LeftJoinLateralExpression : JoinExpressionBase
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftJoinLateralExpression.cs]: Why is this needed?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs [hunk 2]
```diff
@@ -161,34 +161,14 @@ public virtual bool IsIdentityQuery()
...
-                tableExpression.Alias = _queryCompilationContext.CreateUniqueTableAlias(tableExpression.Alias);
-            }
             _tables.Add(tableExpression);
         }
 
         /// <summary>
-        ///     Adds tables to this SelectExprssion.
-        /// </summary>
-        /// <param name="tableExpressions"> The table expressions. </param>
-        public virtual void AddTables([NotNull] IEnumerable<TableExpressionBase> tableExpressions)
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs]: Keep this method but make it private.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs [hunk 3]
```diff
@@ -518,6 +514,11 @@ public virtual int AddToProjection([NotNull] Expression expression, bool resetPr
         {
             Check.NotNull(expression, nameof(expression));
 
+            if (expression == this)
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs]: remove this.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 4]
```diff
@@ -722,8 +723,11 @@ var projection
 
                         var joinExpression
                             = !groupJoin
-                                ? previousSelectExpression.AddInnerJoin(tableExpression, projection)
-                                : previousSelectExpression.AddLeftOuterJoin(tableExpression, projection);
+                                ? new InnerJoinExpression(tableExpression)
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: What is the reason to remove methods in favour of new operator?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs (hunk 2 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs]: :+1:

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs (hunk 3 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs]: :+1: yeah, I didn't mean to leave that in there

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftJoinLateralExpression.cs (hunk 1 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftJoinLateralExpression.cs]: It is used for OUTER APPLY/LEFT JOIN LATERAL in the other big PR I am splitting apart. I felt it made sense for this PR because introducing it is what drove the rename of the LateralJoinExpression, and addressing this change now would eliminate some noise later when dealing with the changes that actually use the LeftLateralJoinExpression.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs (hunk 4 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: The methods didn't do anything special at all, and with the addition of LeftJoinLateralExpression, it would be adding yet another method or two to be consistent, or just clean it up a bit by ditching them. SelectExpression has a lot going on and these methods just seem to clutter up the API.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftJoinLateralExpression.cs (hunk 1 above)
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/LeftJoinLateralExpression.cs]: Add the types when you really need them.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs (hunk 4 above)
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: While method may seem to clutter up at the same time they provide re-usability. Also it encapsulate the formation of join inside SelectExpression. Which could help in future to achieve referential integrity. Lets keep the methods unless we have compelling reason to remove them.

Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: I see what you did in recent commits regarding these now, fair enough :+1:

**Comment by tuespetre** [comment]
:up::date:  

I have made the requested changes in addition to catching that I forgot to make `CrossJoinExpression` inherit from `JoinExpressionBase`.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs [hunk 5]
```diff
@@ -507,6 +503,20 @@ public virtual int AddToProjection([NotNull] Expression expression)
             => AddToProjection(expression, true);
 
         /// <summary>
+        ///     Adds a range of expressions to the projection.
+        /// </summary>
+        /// <param name="expressions"> The expressions. </param>
+        public virtual void AddToProjection([NotNull] IEnumerable<Expression> expressions)
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Expressions/SelectExpression.cs]: Is this still being referenced?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs [hunk 6]
```diff
@@ -152,22 +152,24 @@ protected override Expression VisitEntityQueryable(Type elementType)
         {
             Check.NotNull(elementType, nameof(elementType));
 
-            var relationalQueryCompilationContext = QueryModelVisitor.QueryCompilationContext;
+            var compilationContext = QueryModelVisitor.QueryCompilationContext;
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs]: Preserve variable name.

**Comment by smitpatel** [comment]
Looks good otherwise 😃

**Comment by tuespetre** [comment]
:up::date:

**Comment by tuespetre** [comment]
:up::date: to rebase onto the VS2017 changes

**Comment by smitpatel** [comment]
Merged via d95988daa3b866b09acbb3e2b457cb5eee0f3c68
